### ISSUE TITLE
`Internal compiler error during compilation` for a function with `external` modifier

### ISSUE THREAD

**Issue by ghost** [ethereum/solidity#3187]
Whenever I do:
```
✗ solc --abi --bin src/EthereumExperiment.sol
```

on the following source code file:
```
pragma solidity ^0.4.18;

contract EthereumExperiment {

    struct Actual {
        uint256 a;
    }

    Actual[] public actuals;

    function registerActual(Actual actual) external {
        actuals.push(actual);
    }

}
```

I get the following error:
```
Internal compiler error during compilation:
/tmp/solidity-20171023-17950-1b1unaw/solidity_0.4.18/libsolidity/codegen/CompilerUtils.cpp(709): Throw in function void dev::solidity::CompilerUtils::convertType(const dev::solidity::Type &, const dev::solidity::Type &, bool, bool, bool)
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what:
[dev::tag_comment*] =
```

My solidity version is:

```
✗ solc --version
solc, the solidity compiler commandline interface
Version: 0.4.18+commit.9cf6e910.Darwin.appleclang
```

Changing the `external` modifier to `public` makes the contract compile successfully.

**Comment by ghost** [comment]
Am I doing something wrong?
Or did I just find a bug in the compiler?

**Comment by federicobond** [comment]
It's indeed a bug in the compiler. The current ABI encoder does not support passing structs from external calls, but it should show a proper error message.

For the moment, you can opt-in to the new encoder with `pragma experimental ABIEncoderV2;` or pass each individual struct member as separate arguments.

**Comment by axic** [comment]
With the latest compiler, without the `ABIEncoderV2` an error is emitted:
```
3187.sol:11:29: Error: Structs are only supported in the new experimental ABI encoder. Use "pragma experimental ABIEncoderV2;" to enable the feature.
    function registerActual(Actual actual) external {
                            ^-----------^
```

With `ABIEncoderV2` it is an unimplemented assertion:
```
Unimplemented feature:
/Users/alex/Projects/solidity/libsolidity/codegen/ABIFunctions.cpp(1265): Throw in function string dev::solidity::ABIFunctions::abiDecodingFunctionStruct(const dev::solidity::StructType &, bool)
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::UnimplementedFeatureError>
std::exception::what: 
[dev::tag_comment*] = 
```

The message here should be improved :)

**Comment by chriseth** [comment]
This has been fixed with #3549
